src/main/java/org/jboss/modules/PathUtils.java

```diff
@@ -274,6 +265,27 @@ public static String canonicalize(String path) {
         return new String(targetBuf, a + 1, length - a - 1);
     }
 
+    /**
+     * Canonicalize the given path.  Removes all {@code .} and {@code ..} segments from the path.
+     *
+     * @param path the relative or absolute possibly non-canonical path
+     * @return the canonical path
+     */
+    public static String canonicalize(String path) {
+        if (path.trim().isEmpty() || !path.contains(".")) {
```

Comment:
I'd try two things here:

* Drop the `trim()` part. If the path does not consist of only whitespace and does have leading or trailing whitespace, this will cause a new `String` allocation which is immediately discarded.
* Try changing `!path.contains(".")` to `path.indexOf('.') == -1`, which should eliminate a nested loop and some other extra work inside of `String`.